python/ray/dag/tests/experimental/test_collective_dag.py

```diff
@@ -0,0 +1,304 @@
+# coding: utf-8
+import logging
+import os
+import sys
+import uuid
+from typing import Dict, FrozenSet, List, Optional, Set, Tuple
+
+import pytest
+import ray
+import ray.cluster_utils
+import ray.experimental.collective as collective
+import torch
+from ray.dag import InputNode, MultiOutputNode
+from ray.experimental.channel.gpu_communicator import (
+    GPUCommunicator,
+    TorchTensorAllocator,
+)
+from ray.tests.conftest import *  # noqa
+from ray.util.collective.types import ReduceOp
+
+logger = logging.getLogger(__name__)
+
+if sys.platform != "linux" and sys.platform != "darwin":
+    pytest.skip("Skipping, requires Linux or Mac.", allow_module_level=True)
+
+
+@ray.remote
+class CPUTorchTensorWorker:
+    def __init__(self):
+        self.device = "cpu"
+
+    def send(self, shape, dtype, value: int, send_tensor=True):
+        if not send_tensor:
+            return 1
+        return torch.ones(shape, dtype=dtype, device=self.device) * value
+
+    def recv(self, tensor):
+        # Check that tensor got loaded to the correct device.
+        assert tensor.device == self.device
+        return (tensor[0].item(), tensor.shape, tensor.dtype)
+
+    def compute_with_tuple_args(self, args, i: int):
+        shape, dtype, value = args[i]
+        tensor = torch.ones(shape, dtype=dtype, device=self.device) * value
+        return tensor
+
+
+class MockNcclGroupSet:
+    def __init__(self):
+        # Represents a mapping from a NCCL group ID to a set of actors and a custom
+        # NCCL group.
+        self.ids_to_actors_and_custom_comms: Dict[
+            str, Tuple[FrozenSet["ray.actor.ActorHandle"], Optional[GPUCommunicator]]
+        ] = {}
+
+    def __call__(
+        self,
+        actors: List["ray.actor.ActorHandle"],
+        custom_nccl_group: Optional[GPUCommunicator] = None,
+    ) -> str:
+        nccl_group_id = str(uuid.uuid4())
+        self.ids_to_actors_and_custom_comms[nccl_group_id] = (
+            frozenset(actors),
+            custom_nccl_group,
+        )
+        return nccl_group_id
+
+    def check_init(
+        self,
+        compiled_dag: "ray.dag.CompiledDAG",
+        num: int,
```

Comment:
Can remove this and just use `len(actors_and_custom_comms)`

Reply:
Ack.